core/recorder.h

```diff
@@ -0,0 +1,127 @@
+
+    /** Record an expression value as a row.  This will be flattened by
+        datasets that require flattening.
+    */
+    virtual void
+    recordRowExpr(const RowName & rowName,
+                  const ExpressionValue & expr) = 0;
+
+    /** See recordRowExpr().  This has the same effect, but takes an rvalue
+        which is destroyed by the call.  This may result in performance
+        improvements.
+    */
+    virtual void
+    recordRowExprDestructive(RowName rowName,
+                             ExpressionValue expr);
+
+    /** Record a pre-flattened row. */
+    virtual void
+    recordRow(const RowName & rowName,
+              const std::vector<std::tuple<ColumnName, CellValue, Date> > & vals) = 0;
+
+    /** See recordRow().  This has the same effect, but takes an rvalue
+        which is destroyed by the call.  This may result in performance
+        improvements.
+    */
+    virtual void
+    recordRowDestructive(RowName rowName,
+                         std::vector<std::tuple<ColumnName, CellValue, Date> > vals);
+
+    /** Record multiple flattened rows in a single transaction.  Default
+        implementation forwards to recordRow.
+    */
+    virtual void
+    recordRows(const std::vector<std::pair<RowName, std::vector<std::tuple<ColumnName, CellValue, Date> > > > & rows) = 0;
+
+    /** See recordRows().  This has the same effect, but takes an rvalue
+        which is destroyed by the call.  This may result in performance
+        improvements.
+    */
+    virtual void
+    recordRowsDestructive(std::vector<std::pair<RowName, std::vector<std::tuple<ColumnName, CellValue, Date> > > > rows);
+
+    /** Record multiple rows from ExpressionValues.  */
+    virtual void
+    recordRowsExpr(const std::vector<std::pair<RowName, ExpressionValue > > & rows) = 0;
+
+    /** See recordRowsExpr().  This has the same effect, but takes an rvalue
+        which is destroyed by the call.  This may result in performance
+        improvements.
+    */
+    virtual
+    void recordRowsExprDestructive(std::vector<std::pair<RowName, ExpressionValue > > rows);
+
+    /** Return a function specialized to record the same set of atomic values
+        over and over again into this chunk.
+
+        The returned function can be called row by row, without needing to
+        mention column names (inferred by the positions).  The values will
+        be destroyed on insertion.  numVals must always equal the number of
+        columns passed.
+
+        Extra allows for unexpected columns to be recorded, for when the
+        dataset contains some fixed columns but also may have dynamic
+        columns.
+    */
+    virtual
+    std::function<void (RowName rowName,
+                        Date timestamp,
+                        CellValue * vals,
+                        size_t numVals,
+                        std::vector<std::pair<ColumnName, CellValue> > extra)>
+    specializeRecordTabular(const std::vector<ColumnName> & columns);
```

Comment:
It's not a specialization for a specific dataset.  It's a specialization for a particular data layout.  Any dataset that can record could improve performance by implementing a specialization, as it allows us to avoid manipulating column names.